### ISSUE TITLE
More fine-grained multiline-ness

### ISSUE THREAD

**Issue by georgefst** [fourmolu/fourmolu#46]
[This rejected commit](https://github.com/georgefst/monpad/commit/708897cfb5281b3ef65ecce43106bd14e0c41622) demonstrates some of the remaining places in which I find Fourmolu's default style unpalatable.

Essentially, where an expression is only multiline because it ends with a particular *multiline-friendly construct* (this is a bit vague, but definitely includes record update and construction. `case` and `\case` statements, multi-way-if, and probably list literals), I'd like to format the preceding part of the expression in single-line style.

As an extreme case, we currently turn:

```hs
f = g 1 2 3 4 5 R
    { a = ()
    , b = ()
    }
```

in to:

```hs

f =
    g
        1
        2
        3
        4
        5
        R
            { a = ()
            , b = ()
            }
```

Which is unnecessarily indented (twice!), and has a lot of almost-empty lines. The only current sensible solution is to use a local binding. And that only solves the second issue.

Unfortunately I'm not entirely convinced this option could be implemented simply or efficiently.

## Another example

It rather annoys me that, while this is accepted, adding some leading operation can cause a whole block below to be indented:
```hs
    items <- for [1 .. 5] \i ->
        print i
```
```hs
    items <-
        id <$> for [1 .. 5] \i -> -- note that `$` is special-cased - that wouldn't be reformatted
            print i
```

**Comment by georgefst** [comment]
Okay, actually, a `$` before the `R` largely solves the excess vertical space. Similar use of `$` works for all similar examples I can think of. But I can't see a way around the excess horizontal space.

**Comment by georgefst** [comment]
> But I can't see a way around the excess horizontal space.

Some speak of a cure for these sorts of ills. But we would not contemplate such heresy...

**Comment by brandonchinn178** [comment]
This is an Ormolu issue: https://github.com/tweag/ormolu/issues/692. It's certainly exacerbated if one uses a bigger indentation size than 2, but Fourmolu doesn't, and IMO shouldn't, do anything different here

**Comment by georgefst** [comment]
> Fourmolu doesn't, and IMO shouldn't, do anything different here

Why do you think that? I'm not so sure. We have the opportunity to do something better downstream, as long as it isn't too messy to implement.

In general, I don't think we should close issues like this, unless we at least have confirmation that upstream actually consider it an issue. Which I suspect will unfortunately not be the case here. The same could perhaps be said for #168.

**Comment by brandonchinn178** [comment]
If Ormolu closes the issue upstream, I'm happy to reopen this. I can't imagine why Ormolu wouldnt want to fix this though. If we figured out a fix and made a PR upstream, why wouldn't ormolu just accept it? And if ormolu closes it, you already have the work, so you can just cherry pick the changes to fourmolu.

**Comment by georgefst** [comment]
> I can't imagine why Ormolu wouldnt want to fix this though.

They're just generally past the point of making big style changes*, so I assume anything remaining that annoys me doesn't annoy any of the Ormolu maintainers nearly as much. https://github.com/tweag/ormolu/issues/692 being open for so long without a response implies that they at least don't see it as a priority.

\* This change at least wouldn't reformat existing code, so it is possible it'd be palatable.

**Comment by georgefst** [comment]
Thanks for linking https://github.com/tweag/ormolu/issues/692 anyway, as I hadn't seen it until now. And while this issue is slightly more general, you've explained the problem more clearly than I managed.

**Comment by brandonchinn178** [comment]
Yeah, I agree. But I'd like to push as much of the conversation over to Ormolu as much as possible until Ormolu explicitly says they don't want to do it. At the very least, it would show Ormolu the interest in a particular feature (as opposed to people interacting in two separate discussions).

If we find that these kinds of issues turn people off of Fourmolu, we can put more of a priority on it and start working on it, but there's not much difference implementing it in the ormolu codebase vs ours

**Comment by georgefst** [comment]
I'm reopening since I still care about this a lot, I know from various conversations that I'm not the only one, and it's clear at this point that upstream aren't interested.

I'll try to look in to this again at some point soon, although I do still have some concerns about implementation difficulty.

**Comment by brandonchinn178** [comment]
Pinged mrkkrp on the Ormolu thread to see if they have any more comments. It'd be nice to avoid diverging, as it would make merging upstream more annoying if this adds a lot of conflicts
